plasmapy/classes/sources/openpmd_hdf5.py

```diff
@@ -0,0 +1,136 @@
+
+# This is the order what OpenPMD uses to store unit
+# dimensions for a record.
+_UNITS = (u.meter,
+          u.kilogram,
+          u.second,
+          u.ampere,
+          u.Kelvin,
+          u.mol,
+          u.candela)
+
+
+def _fetch_units(openPMD_dims):
+    """
+    Converts a collection of OpenPMD dimensions to astropy.units.
+    """
+
+    units = u.dimensionless_unscaled
+    for factor, unit in zip(openPMD_dims, _UNITS):
+        units *= (unit ** factor)
+    units, *_ = units.compose()
+    return units
+
+
+def _valid_version(openPMD_version,
+                   outdated=_OUTDATED_VERSION, newer=_NEWER_VERSION):
+    """
+    Checks if the passed version is supported or not.
+    """
+
+    parsed_version = StrictVersion(openPMD_version)
+    outdated_version = StrictVersion(outdated)
+    newer_version = StrictVersion(newer)
+    return outdated_version <= parsed_version < newer_version
+
+
+class HDF5Reader(GenericPlasma):
+    def __init__(self, hdf5, **kwargs):
+        """
+        Core class for accessing various attributes on HDF5 files that
+        are based on OpenPMD standards.
+
+        Attributes
+        ----------
+        electric_field : `astropy.units.Quantity`
+            An (x, y, z) array containing electric field data.
+        charge_density : `astropy.units.Quantity`
+            An array containing charge density data.
+
+        Parameters
+        ----------
+        hdf5 : `str`
+            Path to HDF5 file.
+        """
+
+        if not os.path.isfile(hdf5):
+            raise FileNotFoundError(f"Could not find file: '{hdf5}'")
+
+        h5 = h5py.File(hdf5)
+        self.h5 = h5
+
+        self._check_valid_openpmd_version()
+
+        self.subname = tuple(self.h5['data'])[0]
+
+    def _check_valid_openpmd_version(self):
+        try:
+            openPMD_version = self.h5.attrs["openPMD"].decode('utf-8')
+            if _valid_version(openPMD_version):
+                return True
+            else:
+                raise OpenPMDError(f"We currently only support HDF5 versions"
+                                   f"starting from v{_OUTDATED_VERSION} and "
+                                   f"lower than v{_NEWER_VERSION}. You can "
+                                   f"however convert your HDF5 to a supported "
+                                   f"version. For more information; see "
+                                   f"https://github.com/openPMD/openPMD-updater")
+        except KeyError:
+            raise OpenPMDError("Input HDF5 file does not go on with "
+                               "standards defined by OpenPMD")
+
+    @property
+    def electric_field(self):
+        path = 'data/' + self.subname + '/fields/E'
+        if path in self.h5:
+            units = _fetch_units(self.h5[path].attrs["unitDimension"])
+            return np.array((self.h5[path]['x'],
+                             self.h5[path]['y'],
+                             self.h5[path]['z'])) * units
```

Comment:
I'll definitely have to look at how `np.array` here is going to work... but it's nice enough for the prototype!